executor/write.go

```diff
@@ -154,7 +154,13 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 		if err != nil {
 			return false, handleChanged, newHandle, 0, errors.Trace(err)
 		}
-		newHandle, err = t.AddRecord(ctx, newData, skipHandleCheck)
+		newHandle, err = t.AddRecord(ctx, newData, skipHandleCheck, []bool{true}...)
+		if err != nil {
+			return false, handleChanged, newHandle, 0, errors.Trace(err)
+		}
+		if onDup {
+			sc.AddAffectedRows(1)
+		}
```

Comment:
is this cherry-pick by #8233 & #6656？

maybe better cherry their testcase to here too, look good to me but cannot confirm it works well without testcase... :rofl:

Reply:
@crazycs520 Please pick the test cases.

Reply:
I have cherry-picked test of #6656, But I can't cherry-pick the test of #8233, There have some other problems.

Reply:
I add a part of tests from #8233, PTAL. @lysu